src/controllers/sharedExecutionPlanUtils.ts

```diff
@@ -0,0 +1,184 @@
+        language: "xml",
+    });
+
+    void vscode.window.showTextDocument(planXmlDoc);
+
+    return state;
+}
+
+export async function showQuery(
+    state: QueryResultWebviewState | ExecutionPlanWebviewState,
+    payload: ExecutionPlanReducers["showQuery"],
+    untitledSqlDocumentService: UntitledSqlDocumentService,
+) {
+    void untitledSqlDocumentService.newQuery(payload.query);
+
+    return state;
+}
+
+export async function updateTotalCost(
+    state: QueryResultWebviewState | ExecutionPlanWebviewState,
+    payload: ExecutionPlanReducers["updateTotalCost"],
+) {
+    return {
+        ...state,
+        executionPlanState: {
+            ...state.executionPlanState,
+            totalCost: (state.executionPlanState.totalCost +=
+                payload.addedCost),
+        },
+    };
+}
+
+export async function createExecutionPlanGraphs(
+    state: QueryResultWebviewState | ExecutionPlanWebviewState,
+    executionPlanService: ExecutionPlanService,
+    xmlPlans: string[],
+) {
+    let newState = {
+        ...state.executionPlanState,
+    };
+    const startTime = performance.now(); // timer for telemetry
+    for (const plan of xmlPlans) {
+        const planFile: ExecutionPlanGraphInfo = {
+            graphFileContent: plan,
+            graphFileType: ".sqlplan",
+        };
+        try {
+            newState.executionPlanGraphs = newState.executionPlanGraphs.concat(
+                (await executionPlanService.getExecutionPlan(planFile)).graphs,
+            );
+
+            newState.loadState = ApiStatus.Loaded;
+
+            sendActionEvent(
+                TelemetryViews.QueryPlan,
+                TelemetryActions.OpenQueryPlan,
+                {},
+                {
+                    numberOfPlans:
+                        state.executionPlanState.executionPlanGraphs.length,
+                    LoadTimeInMs: performance.now() - startTime,
+                },
+            );
+        } catch (e) {
+            newState.loadState = ApiStatus.Error;
+            newState.errorMessage = e.toString();
+            sendErrorEvent(
+                TelemetryViews.QueryPlan,
+                TelemetryActions.OpenQueryPlan,
+                e,
+                true, // includeErrorMessage
+            );
+        }
+    }
+    state.executionPlanState = newState;
+    state.executionPlanState.totalCost = calculateTotalCost(state);
+
+    return state;
+}
+
+export function calculateTotalCost(
+    state: QueryResultWebviewState | ExecutionPlanWebviewState,
+): number {
+    if (!state.executionPlanState.executionPlanGraphs) {
+        state.executionPlanState.loadState = ApiStatus.Error;
+        return 0;
+    }
+
+    let sum = 0;
+    for (const graph of state.executionPlanState.executionPlanGraphs) {
+        sum += graph.root.cost + graph.root.subTreeCost;
+    }
+    return sum;
+}
+
+export function formatXml(xmlContents: string): string {
+    let formattedXml = "";

```

Comment:
Did you get this logic from somewhere?  If there's a package library that does XML formatting, that might be a less-risky option than custom code.

Perhaps it'd be a good idea to wrap it all in a try-catch where catch just returns the original XML just in case.

Reply:
i didn't get the code from somewhere, but i'll wrap it in a try catch and investigate packages to format it